MantidPlot/src/WindowFactory.h

```diff
@@ -0,0 +1,203 @@
+   * @param app :: handle to the ApplicationWindow instance
+   * @param fileVersion :: the file version number
+   * @return an instance of IProjectSerialisable
+   */
+  IProjectSerialisable *loadFromProject(const std::string &className,
+                                        const std::string &lines,
+                                        ApplicationWindow *app,
+                                        const int fileVersion = -1) {
+    auto it = _map.find(className);
+    if (it != _map.end())
+      return it->second->loadFromProject(lines, app, fileVersion);
+    else
+      throw Kernel::Exception::NotFoundError(
+          "WindowFactory: " + className + " is not registered.\n", className);
+  }
+
+  /** Subscribe a class to the factory
+   *
+   * @param className :: the name of the class to subscribe
+   */
+  template <class C> void subscribe(const std::string &className) {
+    auto instantiator = Mantid::Kernel::make_unique<
+        ProjectWindowInstantiator<C, IProjectSerialisable>>();
+    subscribe(className, std::move(instantiator));
+  }
+
+  /** Get all of the keys for the classes subscribed to the factory
+   *
+   * @return a vector of strings for classes subscribed to the factory.
+   */
+  virtual const std::vector<std::string> getKeys() const {
+    std::vector<std::string> names;
+    names.reserve(_map.size());
+    std::transform(
+        _map.cbegin(), _map.cend(), std::back_inserter(names),
+        [](const std::pair<const std::string,
+                           std::unique_ptr<AbstractFactory>> &mapPair) {
+          return mapPair.first;
+        });
+    return names;
+  }
+
+private:
+  /** Subscribe a class to the factory.
+   *
+   * @param className :: the name of the class to subscribe
+   * @param instantiator :: the instantiator object to used to create the object
+   */
+  void subscribe(const std::string &className,
+                 std::unique_ptr<AbstractFactory> instantiator) {
+    if (className.empty()) {
+      throw std::invalid_argument("Cannot register empty class name");
+    }
+
+    auto it = _map.find(className);
+    if (it == _map.end()) {
+      _map[className] = std::move(instantiator);
+    } else {
+      throw std::runtime_error(className + " is already registered.\n");
+    }
+  }
+
+  /// A typedef for the map of registered classes
+  typedef std::map<std::string, std::unique_ptr<AbstractFactory>,
+                   CaseInsensitiveStringComparator> FactoryMap;
+  /// The map holding the registered class names and their instantiators
+  FactoryMap _map;
+};
+
+/// Forward declaration of a specialisation of SingletonHolder for
+/// AlgorithmFactoryImpl (needed for dllexport/dllimport) and a typedef for it.
+#ifdef _WIN32
+// this breaks new namespace declaraion rules; need to find a better fix
+template class Mantid::Kernel::SingletonHolder<WindowFactoryImpl>;
+#endif /* _WIN32 */
+
+typedef Mantid::Kernel::SingletonHolder<WindowFactoryImpl> WindowFactory;
+}
```

Comment:
You're correct that the symbol exporting doesn't apply to MantidPlot but this code suggests, with the comments etc that it is involved in that mechanism somehow.

I think we should clean up the code at the bottom here so that it more closely follows that as defined in the link provided by @AntonPiccardoSelg. At the very least the `SingletonHolder` template needs to be specialized in the Mantid::Kernel namespace. I think all we need to do is remove the `EXTERN_MANTID_API` from the example.